src/pages/index.js

```diff
@@ -63,27 +100,102 @@ function handleImageClick(cardLink, cardCaption) {
   imagePopup.open(cardLink, cardCaption);
 }
 
+function handleAvatarClick() {
+  editAvatarPopup.open();
+}
+
+function handleDeletePopup(cardId, evt) {
+  deleteVerifyPopup.open();
+  cardIdForDeletion = cardId;
+  cardToDelete = evt.target.closest(".card");
+}
+
+function removeLikeFromServer(cardId, likeButton) {
+  api
+    .removeLikeFromCard(cardId)
+    .then(likeButton.classList.remove("card__liked_active"))
+    .catch((err) => console.error(err));
+}
+
+function likeToServer(cardId, likeButton) {
+  api
+    .addLikeToCard(cardId)
+    .then(likeButton.classList.add("card__liked_active"))
+    .catch((err) => console.error(err));
+}
+
 function createCard(imageInfo) {
-  const newCard = new Card(imageInfo, "#card-template", handleImageClick);
+  const newCard = new Card(
+    imageInfo,
+    "#card-template",
+    handleImageClick,
+    handleDeletePopup,
+    removeLikeFromServer,
+    likeToServer
+  );
   const cardElement = newCard.generateCard();
   return cardElement;
 }
 
+function handleSubmit(request, popupInstance, loadingText = "Saving...") {
+  popupInstance.renderLoading(true, loadingText);
+  request()
+    .then(() => {
+      popupInstance.close();
+    })
+    .catch(console.error)
+    .finally(() => {
+      popupInstance.renderLoading(false);
+    });
+}
+
 function handleAddImage(imageInfo) {
-  const addedCardElement = createCard(imageInfo);
-  cardSection.addItem(addedCardElement, "prepend");
-  addFormPopup.close();
+  function makeRequest() {
+    return api.addCard(imageInfo).then((res) => {
+      const addedCardElement = createCard(res);
+      cardSection.addItem(addedCardElement, "prepend");
+    });
+  }
+  handleSubmit(makeRequest, addFormPopup);
   addFormPopup.clearForm();
   formValidators["add"].disableSubmitButton();
 }
 
 function handleEditProfile(newProfileInfo) {
-  userInfo.setUserInfo(newProfileInfo);
-  editFormPopup.close();
+  function makeRequest() {
+    return api
+      .editUserInfo(newProfileInfo)
+      .then(userInfo.changeUserInfo(newProfileInfo))
+      .then(userInfo.setUserInfo(profileInfo));
+  }
+  handleSubmit(makeRequest, editFormPopup);
+}
+
+function handleChangeAvatar(newAvatarImage) {
+  function makeRequest() {
+    return api
+      .changeProfilePicture(newAvatarImage)
+      .then(userInfo.changeAvatar(newAvatarImage.avatar))
+      .then(userInfo.setUserInfo(profileInfo));
+  }
+  handleSubmit(makeRequest, editAvatarPopup);
+  editAvatarPopup.clearForm();
+  formValidators["avatar"].disableSubmitButton();
+}
+
+function handleDeleteVerify() {
+  api
+    .deleteCard(cardIdForDeletion)
+    .then(cardToDelete.remove())
+    .catch((err) => console.error(err))
+    .finally(deleteVerifyPopup.close());
```

Comment:
You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well